packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+- Using `children` vs custom render props vs arbitrary "data" props
+- Controlled and semi-controlled components
+- Composition patterns
+- Sub-components naming conventions
+- Components' layout responsibilities and boundaries (i.e., a component should only affect the layout of its children, not its own)
+- ...
+
+### APIs Consinstency
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`) or `has*` (`hasValue`).
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- `styled` components should be suffixed with `*View` (e.g. `BoxView`)
+- ...
+
+### Technical requirements for new components
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group);
+
+#### TypeScript
+
+All new components should be written in TypeScript and should be typed using the `PolymorphicComponent` type (more details about polymorphism can be found above in the "Components composition" section).
+
+#### Emotion
+
+All new component should be styled using [Emotion](https://emotion.sh/docs/introduction).
+
+Note: Instead of using Emotion's standard `cx` function, the custom [`useCx` hook](/packages/components/src/utils/hooks/use-cx.ts) should be used instead.
+
+#### Context system
+
+[To be expanded]
+
+#### Hooks vs Components
+
+[To be expanded]
+
+#### Unit tests
+
+All new components should include unit tests using [testing library](https://testing-library.com/). Ideally, the component's main piece of functionality would be tested explicitely (rather than using snapshots).
+
+#### Storybook
+
+All new components should include Storybook examples.
+
+#### Documentation
+
+All components, in addition to being typed, should be using JSDoc when necessary.
+
+Each component that is exported from the `@wordpress/components` package should include a `README.md` file, explaining how to use the component, showing examples, and documenting all the props.
+
+#### Folder structure
+
+Following all previous guidelines, all new components should follow this folder structure:
+
+```
+component-parent-folder/
+├── sub-component-folder/
+├── component.tsx
+├── context.ts
+├── hook.ts
+├── index.ts
+├── README.md
+├── styles.ts
+└── types.ts
+```
+
+In case of a family of components (e.g. `Card` and `CardBody`, `CardFooter`, `CardHeader` ...), each component's implementation should be added in a separate subfolder:
+
+```
+component-parent-folder/
+├── sub-component-folder/
+│   ├── index.ts
+│   ├── component.tsx
+│   ├── hook.ts
+│   └── README.md
+├── sub-component-folder/
+│   ├── index.ts
+│   ├── component.tsx
+│   ├── hook.ts
+│   └── README.md
+├── stories
+│   └── index.js
+├── test
+│   └── index.js
+├── context.ts
+├── index.ts
+├── styles.ts
+└── types.ts
```

Comment:
It always weirded me out that we split components into separate folders but cram all the types and styles into single files... why is that? Why not split the styles along with the components that use them?

Reply:
That's what I have always thought too — why shouldn't each component have its own `styles.ts` and `types.ts` types?

I can update the guidelines to reflect this.